### PR TITLE
Fixes build for PyPy3

### PR DESCRIPTION
**Pull request by mjhoffman65**
@auvipy, I took at stab at fixing the PyPy failures.  My first step is to try to install all the requirements and recreate the unit test failure in the docker environment.  I ran into issues installing the python requirements and am wondering if more work needs to be done before we can expect celery to run with PyPy.  Looks like at least a couple of the dependencies do not work with PyPy since they rely on a C backend.

couchbase: https://docs.couchbase.com/python-sdk/current/hello-world/start-using-sdk.html#pypy-support
ephem: https://github.com/celery/django-celery-beat/issues/69#issue-247235755 (Note we may be able to replace this with skyfield.  See [here](https://rhodesmill.org/pyephem/) for more.)

This PR illustrates the same error I'm getting in the docker environment: https://github.com/celery/celery/runs/1930241932?check_suite_focus=true

How should I move forward?

### CODE REVIEW

**Comment by auvipy** [comment]
I think we should stop couchbase testing and eventually remove it from the build. I am not sure how much popular it is.

**Comment by thedrow** [comment]
> 
> 
> I think we should stop couchbase testing and eventually remove it from the build. I am not sure how much popular it is.

Since we're getting PRs for it, it is in use.

**Comment by thedrow** [comment]
What we can do is specify that couchebase cannot be installed in pypy using markers.

**Comment by thedrow** [comment]
We now see actual errors.
Are you working on fixing those?

**Comment by mjhoffman65** [comment]
> What we can do is specify that couchebase cannot be installed in pypy using markers.



> We now see actual errors.
> Are you working on fixing those?



> We now see actual errors.
> Are you working on fixing those?
Yep!  Working now to fix them.

**Comment by mjhoffman65** [comment]
@auvipy @thedrow This is ready for review.  I have some in-progress Dockerfile changes so it's easy to run the unit tests with pypy3 via docker as well, but that can go separately.

File: t/unit/backends/test_mongodb.py [hunk 1]
```diff
@@ -45,7 +45,6 @@ def setup(self):
         self.patching('celery.backends.mongodb.MongoBackend.encode')
         self.patching('celery.backends.mongodb.MongoBackend.decode')
         self.patching('celery.backends.mongodb.Binary')
-        self.patching('datetime.datetime')
```
Comment (thedrow) [inline t/unit/backends/test_mongodb.py:48]: Why did you remove this?

**Comment by thedrow** [comment]
What about that event is a boolean failure that pops up every now and then?
@auvipy You know what I'm talking about right? Can you find a build log where that happens?
@mjhoffman65 If you repeatedly run the test suite multiple times, does it always pass?

File: t/unit/backends/test_mongodb.py (hunk 1 above)
Comment (mjhoffman65) [reply t/unit/backends/test_mongodb.py:48]: Removing that line fixes this error when running the MongoDB backend tests.
image

**Comment by mjhoffman65** [comment]
> What about that event is a boolean failure that pops up every now and then?
> @auvipy You know what I'm talking about right? Can you find a build log where that happens?
> @mjhoffman65 If you repeatedly run the test suite multiple times, does it always pass?

I've ran it 4 times in docker and so far it has always passed.  I'm not sure how to trigger reruns with the github actions, but I haven't seen that error pop up yet.  I'm assuming it's this one: https://github.com/celery/celery/issues/6489

File: t/unit/backends/test_mongodb.py (hunk 1 above)
Comment (thedrow) [reply t/unit/backends/test_mongodb.py:48]: Why does it pass on CPython?

**Comment by thedrow** [comment]
Yes, exactly.

**Comment by mjhoffman65** [comment]
Was able to recreate #6489.  Will reopen after I fix it.

**Comment by thedrow** [comment]
> Was able to recreate #6489. Will reopen after I fix it.

There's no need to close this PR. Just mark it as a draft if it's not ready for merge.

File: celery/utils/timer2.py [hunk 2]
```diff
@@ -93,8 +93,9 @@ def run(self):
     def stop(self):
         self._is_shutdown.set()
         if self.running:
-            self._is_stopped.wait()
-            self.join(THREAD_TIMEOUT_MAX)
+            if not self._is_stopped:
```
Comment (thedrow) [inline celery/utils/timer2.py:96]: Why is this a boolean on PyPy?

**Comment by thedrow** [comment]
It seems like the failing test is not related to your change.
Would you mind investigating or should I open an issue?

File: celery/utils/timer2.py (hunk 2 above)
Comment (thedrow) [reply celery/utils/timer2.py:96]: @mjhoffman65 Ping?

**Comment by thedrow** [comment]
I've rebased this PR but I'm still waiting for your response @mjhoffman65.

**Comment by thedrow** [comment]
@celery/core-developers Can anyone complete this PR?

**Comment by maybe-sybr** [comment]
Those dockerfile changes @mjhoffman65 mentioned might make life easier, but I don't see a ref to them. In any case, I've got a pypy3 interpreter on my box so I'll see if I can clean this up.

File: celery/utils/timer2.py (hunk 2 above)
Comment (maybe-sybr) [reply celery/utils/timer2.py:96]: I'm suspicious about this change. If this is falsey, we're going to try an `.wait()` on it which seems odd since that probably implies that it's some garbage object. I don't think `Event`s should be treated like this, and if it is a `boolean` in the `pypy3` case, then we're going to do `bool.wait()` which seems wrong.

**Comment by maybe-sybr** [comment]
Rebased this to hopefully fix the CI failures on that delivery info test, also with the correct fix for #6489

File: t/unit/backends/test_mongodb.py (hunk 1 above)
Comment (maybe-sybr) [reply t/unit/backends/test_mongodb.py:48]: `tox.ini` appears to need a change to run these tests.
```diff
diff --git a/tox.ini b/tox.ini
index f62ea3cdf..51cf5d020 100644
--- a/tox.ini
+++ b/tox.ini
@@ -31,7 +31,7 @@ deps=
     3.6,3.7,3.8,3.9: -r{toxinidir}/requirements/test-ci-default.txt
     3.5,3.6,3.7,3.8,3.9: -r{toxinidir}/requirements/docs.txt
     3.6,3.7,3.8,3.9: -r{toxinidir}/requirements/docs.txt
-    pypy3: -r{toxinidir}/requirements/test-ci-base.txt
+    pypy3: -r{toxinidir}/requirements/test-ci-default.txt
 
     integration: -r{toxinidir}/requirements/test-integration.txt
 
```

Comment (maybe-sybr) [reply t/unit/backends/test_mongodb.py:48]: This patch was originally added back in the depths of time: 3ea8824d9b It seems to be a vestigial thing, so I think it'll be fine to remove it.

The breakage is because pypy3 appears to be a little less duck-type-y with its datetime subtraction implementation. We're giving it a `Mock` that looks like a `datetime` object, but isn't probably isn't using a `spec=` to pass `isinstance()` checks.

Comment (maybe-sybr) [reply t/unit/backends/test_mongodb.py:48]: Leaving this for @thedrow to resolve

**Comment by thedrow** [comment]
Why is the build failing?

**Comment by maybe-sybr** [comment]
It's still broken :( I de-drafted this thinking it would run through and go green and now can't work out how to put it back :grimacing:

**Comment by maybe-sybr** [comment]
Freshened this on top of master now that 5.1 is released. I've also made some changes to the workflow to run unit tests using tox. I think that might fix up some of the weirdness around the missing dependencies. We'll see...

**Comment by maybe-sybr** [comment]
Freshened on master and also with a few extra fixes for the same kind of `Thread._is_stopped` shadowing in the `bgThread` class which is probably broken in a race-y way similar to `Timer` but only touched in two tests so we probably never noticed that it was different to the timer failure.

**Comment by maybe-sybr** [comment]
Huh, weird, it's trying to run flake8 under action environments without python 3.9 installed :/

**Review by auvipy — APPROVED** [review]

**Review by thedrow — APPROVED** [review]

**Comment by thedrow** [comment]
Thanks to everyone who helped us stablize the build :)
